### PR TITLE
query: deduplicate series

### PR DESCRIPTION
**Pull request by fabxc**
@Bplotka 

Just hacked the SeriesIterator right now, will need to add more testing tomorrow. Overhead is in the range of 25-35ns/sample depending on how "scrambled" the series are. That's below the decoding effort of the underlying chunks and thus should not have a massive impact.

### CODE REVIEW

File: pkg/query/fanin.go [hunk 1]
```diff
@@ -1,188 +0,0 @@
-package query
```
Comment (Bplotka) [inline pkg/query/fanin.go]: Why you moved that around? Easier to navigate through more, smaller files than single big one

File: pkg/query/querier.go [hunk 2]
```diff
@@ -237,3 +247,437 @@ func (q *querier) Close() error {
...
+
+type dedupSeriesSet struct {
+	set          storage.SeriesSet
+	replicaLabel string
+
+	// Gathered replica series so far.
+	replicas []storage.Series
+	lset     labels.Labels
+	peek     storage.Series
+	// lastReplica string
```
Comment (Bplotka) [inline pkg/query/querier.go]: commented var?

File: pkg/query/querier.go [hunk 3]
```diff
@@ -237,3 +247,437 @@ func (q *querier) Close() error {
...
+
+func (s *dedupSeriesSet) Next() bool {
+	if !s.ok {
+		return false
+	}
+	// Set the label set we are currently gathering to the peek element
+	// without the replica label if it exists.
+	s.lset = s.peekLset()
+	s.replicas = append(s.replicas[:0], s.peek)
+	s.peek = nil
```
Comment (Bplotka) [inline pkg/query/querier.go]: why you nil the s.peek here but not before recursive next? (line 533)?

File: pkg/query/querier.go [hunk 4]
```diff
@@ -237,3 +247,437 @@ func (q *querier) Close() error {
...
+	i    int
+
+	aok, bok   bool
+	lastT      int64
+	penA, penB int64
+	useA       bool
+}
+
+func newDedupSeriesIterator(a, b storage.SeriesIterator) *dedupSeriesIterator {
+	it := &dedupSeriesIterator{
```
Comment (Bplotka) [inline pkg/query/querier.go]: just `return &dedupSeriesIterator...`

File: pkg/query/querier.go [hunk 5]
```diff
@@ -237,3 +247,437 @@ func (q *querier) Close() error {
...
+		it.aok = it.a.Seek(it.lastT + 1 + it.penA)
+	}
+	if it.bok {
+		it.bok = it.b.Seek(it.lastT + 1 + it.penB)
+	}
+	// Handle basic cases where one iterator is exhausted before the other.
+	if !it.aok {
+		it.useA = false
+		if it.bok {
+			// it.lastT0 = it.lastT1
```
Comment (Bplotka) [inline pkg/query/querier.go]: commented code

File: pkg/query/querier.go [hunk 6]
```diff
@@ -237,3 +247,437 @@ func (q *querier) Close() error {
...
+			it.lastT, _ = it.b.At()
+			it.penB = 0
+		}
+		return it.bok
+	}
+	if !it.bok {
+		it.useA = true
+		// it.lastT0 = it.lastT1
```
Comment (Bplotka) [inline pkg/query/querier.go]: commented code

File: pkg/query/querier.go [hunk 7]
```diff
@@ -237,3 +247,437 @@ func (q *querier) Close() error {
...
+	// This ensures that we don't pick a sample to close, which would increase the overall
+	// sample frequency. It also guards against clock drift and inaccuracies in timestamp
+	// during timestamp assignment.
+	// If we don't know a delta yet, we pick 3000 as a constant, which is based on the knowledge
+	// that timestamps are in milliseconds and sampling frequencies typically multiple seconds long.
+	if it.useA {
+		if it.lastT != math.MinInt64 {
+			it.penB = ta - it.lastT
+		} else {
+			it.penB = 3000
```
Comment (Bplotka) [inline pkg/query/querier.go]: Let's maybe moved it to constant?

File: pkg/query/fanin.go (hunk 1 above)
Comment (fabxc) [reply pkg/query/fanin.go]: We had two weird small files, didn't feel right. Moved the iterator stuff out into its own file now

**Comment by fabxc** [comment]
Addressed comments, added test cases, and cleaned up benchmark.

**Pull request merged**